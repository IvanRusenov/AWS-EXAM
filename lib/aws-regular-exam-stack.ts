import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {LambdaIntegration, RestApi} from "aws-cdk-lib/aws-apigateway";
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Runtime, StartingPosition} from "aws-cdk-lib/aws-lambda";
import {Subscription, SubscriptionProtocol, Topic} from "aws-cdk-lib/aws-sns";
import {AttributeType, BillingMode, StreamViewType, Table} from "aws-cdk-lib/aws-dynamodb";
import {DynamoEventSource} from "aws-cdk-lib/aws-lambda-event-sources";
import * as path from "node:path";

export class AwsRegularExamStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const topic = new Topic(this, "topic");

        new Subscription(this, "subscription", {
            topic: topic,
            protocol: SubscriptionProtocol.EMAIL,
            endpoint: "i_rusenov@abv.bg"
        });

        const table = new Table(this, "order", {
            partitionKey: {
                name: "PK",
                type: AttributeType.STRING
            },
            sortKey: {
                name: "SK",
                type: AttributeType.STRING
            },
            billingMode: BillingMode.PAY_PER_REQUEST,
            timeToLiveAttribute: "ttl",
            stream: StreamViewType.NEW_AND_OLD_IMAGES,
        });

        const func1 = new NodejsFunction(this, "func1", {
            handler: "handler",
            runtime: Runtime.NODEJS_20_X,
            entry: path.join(__dirname, "../src/func1.ts"),
            environment: {
                // permissions: [topic],
                TABLE_NAME: table.tableName,
                TOPIC_ARN: topic.topicArn,
            }
        });

        topic.grantPublish(func1);
        table.grantReadWriteData(func1)

        const deleteFunc = new NodejsFunction(this, "deleteFunc", {
            handler: "lambdas/deleteFunc.handler",
            runtime: Runtime.NODEJS_20_X,
            entry: path.join(__dirname, "../src/deleteFunc.ts"),

            environment: {
                TABLE_NAME: table.tableName,
                TOPIC_ARN: topic.topicArn,
            },

        });

        table.grantReadWriteData(deleteFunc);
        topic.grantPublish(deleteFunc);

        deleteFunc.addEventSource(new DynamoEventSource(table, {
            startingPosition: StartingPosition.TRIM_HORIZON,
        }));

        const api = new RestApi(this, "api");
        const resource = api.root.addResource("order");
        resource.addMethod("POST", new LambdaIntegration(func1, {proxy: true}));

    }
}
