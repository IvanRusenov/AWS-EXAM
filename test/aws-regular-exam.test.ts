import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {AwsRegularExamStack} from "../lib/aws-regular-exam-stack";

test("Snapshot", () => {
    const app = new App();
    const stack = new AwsRegularExamStack(app,"testStack");
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
});
