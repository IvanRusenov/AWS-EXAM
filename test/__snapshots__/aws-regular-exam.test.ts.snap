// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Outputs": {
    "apiEndpoint9349E63C": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "apiC8550315",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "apiDeploymentStageprod896C8101",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "adminRoleC345D70B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "adminRoleDefaultPolicy4FE403EF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "topic69831491",
              },
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "orderE5E79FED",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "adminRoleDefaultPolicy4FE403EF",
        "Roles": [
          {
            "Ref": "adminRoleC345D70B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "apiAccount57E28B43": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "apiC8550315",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "apiCloudWatchRoleAC81D93E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "apiC8550315": {
      "Properties": {
        "Name": "api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "apiCloudWatchRoleAC81D93E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "apiDeployment149F129420d67c9850f5d0502af6295aedb80313": {
      "DependsOn": [
        "apiorderPOST0E1BB75D",
        "apiorderF613A33C",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "apiC8550315",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "apiDeploymentStageprod896C8101": {
      "DependsOn": [
        "apiAccount57E28B43",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "apiDeployment149F129420d67c9850f5d0502af6295aedb80313",
        },
        "RestApiId": {
          "Ref": "apiC8550315",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "apiorderF613A33C": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "apiC8550315",
            "RootResourceId",
          ],
        },
        "PathPart": "order",
        "RestApiId": {
          "Ref": "apiC8550315",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "apiorderPOST0E1BB75D": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "func14444F01C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "apiorderF613A33C",
        },
        "RestApiId": {
          "Ref": "apiC8550315",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "apiorderPOSTApiPermissionTesttestStackapi4CFEC31DPOSTorder480C84AB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "func14444F01C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "apiC8550315",
              },
              "/test-invoke-stage/POST/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "apiorderPOSTApiPermissiontestStackapi4CFEC31DPOSTorder0803B750": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "func14444F01C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "apiC8550315",
              },
              "/",
              {
                "Ref": "apiDeploymentStageprod896C8101",
              },
              "/POST/order",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "deleteFuncFF12B595": {
      "DependsOn": [
        "deleteFuncServiceRoleDefaultPolicyB98FF212",
        "deleteFuncServiceRole5A8AFB6D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "48e5e41e01c8908a8801826593bcaa5ecb8171524a2f736e6a5c560e4f8d342f.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "orderE5E79FED",
            },
            "TOPIC_ARN": {
              "Ref": "topic69831491",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "deleteFuncServiceRole5A8AFB6D",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "deleteFuncServiceRole5A8AFB6D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deleteFuncServiceRoleDefaultPolicyB98FF212": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "orderE5E79FED",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "topic69831491",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "deleteFuncServiceRoleDefaultPolicyB98FF212",
        "Roles": [
          {
            "Ref": "deleteFuncServiceRole5A8AFB6D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "func14444F01C": {
      "DependsOn": [
        "adminRoleDefaultPolicy4FE403EF",
        "adminRoleC345D70B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "fc3da131d879b2d62210e37a4c46e2c4eb0f4e1a0c36ec8cf36e32df841a0f7e.zip",
        },
        "Environment": {
          "Variables": {
            "DELETE_FUNC_ARN": {
              "Fn::GetAtt": [
                "deleteFuncFF12B595",
                "Arn",
              ],
            },
            "ROLE_ADMIN_ARN": {
              "Fn::GetAtt": [
                "adminRoleC345D70B",
                "Arn",
              ],
            },
            "TABLE_NAME": {
              "Ref": "orderE5E79FED",
            },
            "TOPIC_ARN": {
              "Ref": "topic69831491",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "adminRoleC345D70B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "orderE5E79FED": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "subscriptionD5B264BF": {
      "Properties": {
        "Endpoint": "i_rusenov@abv.bg",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "topic69831491",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "topic69831491": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
